doc/stdenv/cross-compilation.chapter.md

```diff
@@ -78,21 +78,46 @@ If both the dependency and depending packages aren't compilers or other machine-
+
+Dependency types describe the relationships that a package has with each of its transitive dependencies.  You could think of attaching one or more dependency types to each of the formal parameters at the top of a package's `.nix` file, as well as to all of *their* formal parameters, and so on.   Triples like `(foo, bar, baz)`, on the other hand, are a property of an instantiated derivation -- you could would attach a triple `(mips-linux, mips-linux, sparc-solaris)` to a `.drv` file in `/nix/store`.
+
+Only nine dependency types matter in practice:
 
 #### Possible dependency types {#possible-dependency-types}
 
-| Dependency’s host platform | Dependency’s target platform |
-|----------------------------|------------------------------|
-| build                      | build                        |
-| build                      | host                         |
-| build                      | target                       |
-| host                       | host                         |
-| host                       | target                       |
-| target                     | target                       |
+| Dependency type | Dependency’s host platform | Dependency’s target platform |
+|-----------------|----------------------------|------------------------------|
+| build → *       | build                      | (none)                       |
+| build → build   | build                      | build                        |
+| build → host    | build                      | host                         |
+| build → target  | build                      | target                       |
+| host → *        | host                       | (none)                       |
+| host → host     | host                       | host                         |
+| host → target   | host                       | target                       |
+| target → *      | target                     | (none)                       |
+| target → target | target                     | target                       |
+
+Let's use `g++` as an example to make this table clearer.  `g++` is a C++ compiler written in C.  Suppose we are building `g++` with a `(build, host, target)` platform triple of `(foo, bar, baz)`.  This means we are using a `foo`-machine to build a copy of `g++` which will run on a `bar`-machine and emit binaries for the `baz`-machine.
+
+* `g++` links against the host platform's `glibc` C library, which is a "host→ *" dependency with a triple of `(bar, bar, *)`.  Since it is a library, not a compiler, it has no "target".
+
+* Since `g++` is written in C, the `gcc` compiler used to compile it is a "build→ host" dependency of `g++` with a triple of `(foo, foo, bar)`.  This compiler runs on the build platform and emits code for the host platform.
+
+* `gcc` links against the build platform's `glibc` C library, which is a "build→ *" dependency with a triple of `(foo, foo, *)`.  Since it is a library, not a compiler, it has no "target".
```

Comment:
> We could even go into more detail and say `(?, foo, *)` because the build platform doesn't matter. Afterwards, then say it is also `foo` by convention.

Thanks, I'll include that in the follow-up PR.

Also, would you be offended if I changed `foo`, `bar`, and `baz` to `B`, `H`, and `T` respectively?  The example is already very complicated and I feel like having to remember that "`foo` is the build, `bar` is the host, and `T` is the target" without some kind of mnemonic is an extra burden on the reader.

Reply:
> If you call https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/booter.nix manually, you can in fact do more complicated bootstraps with more than 3 total platforms.

Cool, I'm still working my way through the fine details of that and the `{#ssec-stdenv-dependencies}` chapter of the manual.  If I come across any documentation quirks there I'll submit a PR, but if so it will certainly be much smaller than this one.

Reply:
> Also, would you be offended if I changed foo, bar, and baz to B, H, and T respectively? The example is already very complicated and I feel like having to remember that "foo is the build, bar is the host, and T is the target" without some kind of mnemonic is an extra burden on the reader.

Mmm so my idea with `foo` `bar` and `baz was that they are (stand-ins for) concrete platforms. Whereas "build", "host" and "target" I really want to make sure people understand are just *local* names from the perspective of a single derivation, *relative* to it.

A lot of build systems fall in the trap of their being a *global* "build", "host" and "target", and then mistakes of various sorts happen.

Reply:
In the slides linked below, I used some humorous concrete platforms: `i386-linux`, `powerpc-illlumos`, `sparc-windows`, `mips-darwin`, `riscv-sel4`.

Perhaps something like that might be better than `foo`, `bar`, `baz` while continuing to not imply there are is "global" build, host, and target?